TestRun - Password Criteria Checker/Program.cs

```diff
@@ -0,0 +1,175 @@
+﻿/** 
+
+This purpose of this program is to have the user input a password to check it for adherenace to the criteria listed under Criteria Declaration
+
+More criteria will be added as the program grows.
+
+**/
+
+// Criteria Declaration
+// This space is used to declare criteria to be followed for each given password
+string criteria1 = "[1] The password must be 9 to 25 digits long";
+string criteria2 = "[2] The password must contain atleast 1 of the following characters : !,@,#,$,%,^,&,*,(,),-,=,_,+";
+string criteria3 = "[3] The password must contain atleast 1 number character";
+string criteria4 = "[4] The password must contain atleast 1 upper case letter";
+
+
+// User Directons
+// This space outputs to console screen, giving the user information regarding which criteria the given password will be tested against. The criteria output will be the same as the criteria 1-4 strings in the above section.
+Console.WriteLine("This application will test a given Password against the following criteria \n");
+Console.WriteLine(criteria1);
+Console.WriteLine(criteria2);
+Console.WriteLine(criteria3);   
+Console.WriteLine(criteria4);   
+
+
+// Request user password
+// This space requests a password from the user to be tested against established criteria
+Console.WriteLine("Please submit password for testing against criteria : \n");
+string userPass = Console.ReadLine();
+
+// Criteria Compliance Counter
+// This space initiates variables to count the number of criteria to be adhered to (criteriaCounter), as well as the number of criteria that has been adhered to following each check (complianceCounter).
+int complianceCounter = 0;
+int criteriaCounter = 0;
+
+// Compliance Checker Header
+// Cosmetic output to Console
+Console.WriteLine("");
+Console.WriteLine("-------------------------------------------");
+Console.WriteLine("         PASSWORD CRITERIA CHECKER         ");
+Console.WriteLine("-------------------------------------------");
+Console.WriteLine("");
+
+// criteria1
+// Begins testing of criteria1.
+// Adds 1 to criteriaCounter, indicating that there is now 1 additional rule to comply to.
+criteriaCounter++;
+
+// initiates variables denoting the minimum and maximum number of characters in the given password.
+int passLengthMin = 9;
+int passLengthMax = 25;
+// initiates an integer that measures the length of the given password.
+int passLength = userPass.Length;
+// intiates a boolean that confirms the length of the given password to be greater than or equal to the minimum, and less than or equal to the maximum.
+bool passLengthCheck = (passLength >= passLengthMin && passLength <= passLengthMax);
+// according to resulting true or false regard the last boolean check, outputs to console.
+// if true, adds 1 to complianceCounter to indicate passing score within this criteria
+if (passLengthCheck == true ) {

```

Comment:
`passLengthCheck` is already a boolean, so you should be able to pass it directly into your conditional without the additional `== true`

```c#
bool passLengthCheck = (passLength >= passLengthMin && passLength <= passLengthMax);

if (passLengthCheck) { 
  // ...
}
```

Reply:
You probably haven't gotten into creating classes yet, but I think having each criteria in its own class would be useful. Here's how I would do it in Ruby.

```ruby
module PasswordCriteria
  class Length
    MINIMUM = 9
    MAXIMUM = 25

    # @example
    #   PasswordCriteria.call("password1234")
    # 
    # @param password [String] The password to check
    # 
    # @return [Boolean] Does the password pass the length criteria
    # 
    def self.call(password)
      length = password.length

      length >= MINIMUM && length <= MAXIMUM
    end
  end
end
```

There's a number of benefits of having lots of smaller classes. One of which is it becomes easier to test. For example, here's an example test file, written in [Rspec](https://rspec.info/). The general rule of thumb is if you're finding your tests are complicated to write, there's probably some simplification and code extraction that can be done. When writing tests, you only need to test what the class is doing, and you don't need to worry about making sure other classes are working as expected (as you should have tests for those other classes too).

```ruby
describe PasswordCriteria::Length do
  subject { described_class.call(password) } # described_class is PasswordCriteria::Length

  let(:password) { "password1234" }

  it { is_expected.to be true }

  context "with too short of a password" do
    let(:password) { "password" } # 8 characters

    it { is_expected.to be false }
  end

  context "with too long of a password" do
    let(:password) { "passwordafsafsaasfsafafsfafasfsafsaafs" } # Lots characters

    it { is_expected.to be false }
  end
end
```

Once you have one class per criteria, you can specify a list of all of the criteria that you want to run.

```ruby
# ruby
passed_criteria, failed_criteria = [PasswordCriteria::Length, PasswordCriteria::Digits].partition do |criteria_class|
  criteria_class.call(password)
end
```

`passed_criteria` will be an array of all criteria that returned `true`, `failed_criteria` will be everything that returned `false`. You can then assume something like `failed_criteria.length == 0` then all of the password criteria returned true.